src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -161,13 +169,64 @@ static unsigned uct_cuda_copy_iface_progress(uct_iface_h tl_iface)
     unsigned max_events = iface->config.max_poll;
     unsigned count;
 
-    count = uct_cuda_copy_progress_event_queue(&iface->outstanding_d2h_cuda_event_q,
+    count = uct_cuda_copy_progress_event_queue(iface, UCT_CUDA_COPY_STREAM_D2H,
                                                max_events);
-    count += uct_cuda_copy_progress_event_queue(&iface->outstanding_h2d_cuda_event_q,
+    count += uct_cuda_copy_progress_event_queue(iface, UCT_CUDA_COPY_STREAM_H2D,
                                                 (max_events - count));
     return count;
 }
 
+#if (__CUDACC_VER_MAJOR__ >= 100000)
+static void CUDA_CB myHostFn(void *cuda_copy_iface)
+#else
+static void CUDA_CB myHostCallback(CUstream hStream,  CUresult status,
+                                   void *cuda_copy_iface)
+#endif
+{
+    uct_cuda_copy_iface_t *iface = cuda_copy_iface;
+
+    ucs_assert(iface->async.event_cb != NULL);
+    /* notify user */
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    iface->async.event_cb(iface->async.event_arg, 0);
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static ucs_status_t uct_cuda_copy_iface_event_fd_arm(uct_iface_h tl_iface,
+                                                    unsigned events)
+{
+    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_copy_iface_t);
+    int i;
+    ucs_status_t status;
+    uct_cuda_copy_event_desc_t *cuda_event;
+
+    for (i = 0; i < UCT_CUDA_COPY_STREAM_LAST; i++) {
+        ucs_queue_for_each(cuda_event, &iface->outstanding_event_q[i], queue) {
+            if (cudaSuccess == cudaEventQuery(cuda_event->event)) {
+                return UCS_ERR_BUSY;
+            }
+        }
+    }
```

Comment:
this is not the right way
need to add helper func to check if **head of the queue** event is ready or not
and use it both in arm, and as ucs_queue_for_each_extract condition in progress